### PR TITLE
Add EspressoTEEVerifier contract and support SGX

### PR DESCRIPTION
**Pull request by Sneh1999**
### This PR:

This PR adds in a few contracts:
`EspressoTEEVerifier` contract: This is the main contract which will call all the other TEE contracts
`EspressoSGXTEEVerifier` contract: This is the SGX contract which will do the SGX verification

### This PR does not:
Implement `EspressoAWSNitroVerifier`

### Key places to review:
- `EspressoTEEVerifier.sol`
- `EspressoSGXTEEVerifier.sol`

### How to test this PR:
forge test

### CODE REVIEW

File: test/EspressoTEEVerifier.t.sol [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        // Get the instance of the DCAP Attestation QuoteVerifier on the Arbitrum Sepolia Rollup
+        vm.startPrank(adminTEE);
+
+        espressoSGXTEEVerifier =
+            new EspressoSGXTEEVerifier(enclaveHash, enclaveSigner, v3QuoteVerifier);
+        espressoTEEVerifier = new EspressoTEEVerifier(espressoSGXTEEVerifier);
+        vm.stopPrank();
+    }
+
+    function testRegisterSigner() public {
```
Comment (zacshowa) [inline test/EspressoTEEVerifier.t.sol:34]: I see that you have tests in the SGXVerifier test file for invalid quotes and such, but shouldn't we also test invalid registration patterns here? At the very least with an invalid tee type being passed in?

File: test/EspressoTEEVerifier.t.sol [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        assertEq(
+            espressoTEEVerifier.registeredEnclaveSigners(
+                bytes32(0x5fc862cb2e7e1f449f36a18b18aca08c20feaed0d411247816c281d596420cbb),
+                IEspressoTEEVerifier.TeeType.SGX
+            ),
+            true
+        );
+    }
+
+    function testSetEspressoSGXTEEVerifier() public {
```
Comment (zacshowa) [inline test/EspressoTEEVerifier.t.sol:79]: I think similarly here we could encode the test to show that non admins cannot set the address for the sgx verifier. I think that would alert future readers that changes that allow those that aren't the admin to set the sgx verifier address is an undesirable characteristic for the contract.

**Review by zacshowa — COMMENTED** [review]
I had a couple comments on the test, but the code looks good to me!

File: src/EspressoSGXTEEVerifier.sol [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    V3QuoteVerifier public quoteVerifier;
+
+    mapping(bytes32 => bool) public registeredEnclaveHash;
+    mapping(bytes32 => bool) public registeredEnclaveSigner;
+    mapping(address => bool) public registeredSigners;
+
+    constructor(bytes32 enclaveHash, bytes32 enclaveSigner, address _quoteVerifier)
+        Ownable(msg.sender)
+    {
+        quoteVerifier = V3QuoteVerifier(_quoteVerifier);
```
Comment (alysiahuggins) [inline src/EspressoSGXTEEVerifier.sol:36]: you can add checks to ensure that the values are valid e.g. 
```solidity
if (_quoteVerifier == address(0)) return InvalidValue()
```

Also is the quoteVerifier meant to be a contract? If so, you can also use `_quoteVerifier.code.length > 0` to determine if it is a contract but this assumes that this contract is already deployed when this contract, `EspressoSGXTEEVerifier` is being deployed.

File: src/EspressoSGXTEEVerifier.sol [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+ *         from automata to verify the quote. Along with some additional verification logic.
+ */
+contract EspressoSGXTEEVerifier is IEspressoSGXTEEVerifier, Ownable2Step {
+    using BytesUtils for bytes;
+
+    // V3QuoteVerififer contract from automata to verify the quote
+    V3QuoteVerifier public quoteVerifier;
+
+    mapping(bytes32 => bool) public registeredEnclaveHash;
+    mapping(bytes32 => bool) public registeredEnclaveSigner;
```
Comment (alysiahuggins) [inline src/EspressoSGXTEEVerifier.sol:30]: just curious, what's the difference between `registeredSigners` and `registeredEnclaveSigners`?

and I suppose the enclave signers is stored as bytes32 because it's a hash and not an ethereum address?

File: src/EspressoSGXTEEVerifier.sol [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        }
+
+        address signer = address(uint160(bytes20(data[:20])));
+
+        // Check if the extracted address is valid
+        if (signer == address(0)) {
+            revert InvalidSignerAddress(); // Custom revert if the address is invalid
+        }
+        // Mark the signer as registered
+        registeredSigners[signer] = true;
```
Comment (alysiahuggins) [inline src/EspressoSGXTEEVerifier.sol:117]: you can add a check that the signer isn't already registered to prevent unneccessary storage updates

File: src/EspressoSGXTEEVerifier.sol [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        @notice Register a new signer by verifying a quote from the TEE
+        @param attestation The attestation from the TEE
+        @param data which the TEE has attested to
+    */
+    function registerSigner(bytes calldata attestation, bytes calldata data) external {
+        // Check that the data length is 20 bytes because an address is 20 bytes
+        if (data.length != 20) {
+            revert InvalidDataLength();
+        }
+        bytes32 paddedSignerAddress = keccak256(data);
```
Comment (alysiahuggins) [inline src/EspressoSGXTEEVerifier.sol:102]: checking my understanding and following up with a question.
so in this function we are adding a signer which is represented as a regular address, it's passed in as bytes because that's the output of the TEE. Why do we need to hash the data again? Is it because the enclave signed `keccak256(data)`?

File: src/EspressoSGXTEEVerifier.sol [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+        return localReport;
+    }
+
+    /*
+        @notice Register a new signer by verifying a quote from the TEE
+        @param attestation The attestation from the TEE
+        @param data which the TEE has attested to
+    */
+    function registerSigner(bytes calldata attestation, bytes calldata data) external {
```
Comment (alysiahuggins) [inline src/EspressoSGXTEEVerifier.sol:97]: just verifying that we intend for this function be callable by anyone

File: test/EspressoSGXTEEVerifier.t.sol [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        espressoSGXTEEVerifier.registerSigner(sampleQuote, data);
+    }
+
+    function testRegisterSignerInvalidAddress() public {
+        vm.startPrank(adminTEE);
+        string memory quotePath = "/test/configs/attestation.bin";
+        string memory inputFile = string.concat(vm.projectRoot(), quotePath);
+        bytes memory sampleQuote = vm.readFileBinary(inputFile);
+
+        address batchPosterAddress = address(0x4C91660a37d613E1Bd278F9Db882Cc5ED2549072);
```
Comment (alysiahuggins) [inline test/EspressoSGXTEEVerifier.t.sol:70]: perhaps add another check with the zero addresss

File: test/EspressoSGXTEEVerifier.t.sol [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        string memory inputFile = string.concat(vm.projectRoot(), quotePath);
+        bytes memory sampleQuote = vm.readFileBinary(inputFile);
+        espressoSGXTEEVerifier.verify(sampleQuote, reportDataHash);
+        vm.stopPrank();
+    }
+
+    /**
+     * Test verify quote reverts if incorrect header is passed
+     */
+    function testVerifyInvalidHeaderInQuote() public {
```
Comment (alysiahuggins) [inline test/EspressoSGXTEEVerifier.t.sol:129]: would it be worthwhile to add a test for verification of zero values?

File: test/EspressoTEEVerifier.t.sol (hunk 1 above)
Comment (Sneh1999) [reply test/EspressoTEEVerifier.t.sol:34]: okay will do!

File: src/EspressoSGXTEEVerifier.sol (hunk 4 above)
Comment (Sneh1999) [reply src/EspressoSGXTEEVerifier.sol:30]: So each batch poster will generate an `ephemeral key` which each TEE will attest to. We decided to do this because otherwise, we add to submit an attestation with each batch. Now the batch poster can generate an `ephemeral key` and sign the batches using that key instead of sending an attestation every time. 

EnclaveSigners is the TEE signer and registeredSigner is the ephemeral key and yes its the hash of the EnclaveSigner key thats stored

File: src/EspressoSGXTEEVerifier.sol (hunk 5 above)
Comment (Sneh1999) [reply src/EspressoSGXTEEVerifier.sol:117]: true, will do that!

File: src/EspressoSGXTEEVerifier.sol (hunk 7 above)
Comment (Sneh1999) [reply src/EspressoSGXTEEVerifier.sol:97]: yes for anyone with a valid attestation

File: src/EspressoSGXTEEVerifier.sol (hunk 6 above)
Comment (Sneh1999) [reply src/EspressoSGXTEEVerifier.sol:102]: Yes, the enclave signed the hash of the the data ( data is the the bytes representation of the address)

File: test/EspressoSGXTEEVerifier.t.sol (hunk 8 above)
Comment (Sneh1999) [reply test/EspressoSGXTEEVerifier.t.sol:70]: true, will add that!

File: test/EspressoSGXTEEVerifier.t.sol (hunk 9 above)
Comment (Sneh1999) [reply test/EspressoSGXTEEVerifier.t.sol:129]: What would zero values mean here?

File: src/EspressoSGXTEEVerifier.sol (hunk 3 above)
Comment (Sneh1999) [reply src/EspressoSGXTEEVerifier.sol:39]: https://github.com/EspressoSystems/espresso-tee-contracts/pull/2/commits/a2bf9e83c0441bd275e4cb10f5bb23b8d440a8ad

File: test/EspressoTEEVerifier.t.sol (hunk 1 above)
Comment (Sneh1999) [reply test/EspressoTEEVerifier.t.sol:35]: https://github.com/EspressoSystems/espresso-tee-contracts/pull/2/commits/a2bf9e83c0441bd275e4cb10f5bb23b8d440a8ad

File: test/EspressoTEEVerifier.t.sol (hunk 2 above)
Comment (Sneh1999) [reply test/EspressoTEEVerifier.t.sol:90]: https://github.com/EspressoSystems/espresso-tee-contracts/pull/2/commits/a2bf9e83c0441bd275e4cb10f5bb23b8d440a8ad

File: src/EspressoSGXTEEVerifier.sol (hunk 5 above)
Comment (Sneh1999) [reply src/EspressoSGXTEEVerifier.sol]: https://github.com/EspressoSystems/espresso-tee-contracts/pull/2/commits/a2bf9e83c0441bd275e4cb10f5bb23b8d440a8ad

File: test/EspressoTEEVerifier.t.sol (hunk 1 above)
Comment (zacshowa) [reply test/EspressoTEEVerifier.t.sol:35]: Hmm that's weird. I know the solidity compiler represents enums as `uint`'s. There isn't way to pass a `uint` to the function or like trick it by casting a `uint` into the enum type that would exist but isn't a currently named member of the enum? Strange :laughing:

Comment (Sneh1999) [reply test/EspressoTEEVerifier.t.sol:35]: hmm let me try!

File: test/EspressoSGXTEEVerifier.t.sol (hunk 8 above)
Comment (alysiahuggins) [reply test/EspressoSGXTEEVerifier.t.sol:71]: ok got it

File: test/EspressoSGXTEEVerifier.t.sol (hunk 9 above)
Comment (alysiahuggins) [reply test/EspressoSGXTEEVerifier.t.sol:140]: bytes calldata rawQuote, bytes32 reportDataHash - either of these are zero? and the verify function should revert then too

File: src/EspressoSGXTEEVerifier.sol (hunk 4 above)
Comment (alysiahuggins) [reply src/EspressoSGXTEEVerifier.sol:30]: ah got it - where can i find the best diagram/description to explain the workflow of this?

**Review by zacshowa — APPROVED** [review]
